# Project 3
### Group member: Jingmin Chen, Yusen Wu, Jiaqi Min

In [227]:
# import csv,numpy,re library
import csv
import numpy as np
import re
#create a empty list a to store data
a = []
with open('tournamentinfo.txt') as csvfile: 
    readCSV = csv.reader(csvfile)
    for row in readCSV:
        a.append(row)
    # print each row as read by the csv.reader function
#print the tournamentinfo.txt file
a

[['-----------------------------------------------------------------------------------------'],
 [' Pair | Player Name                     |Total|Round|Round|Round|Round|Round|Round|Round| '],
 [' Num  | USCF ID / Rtg (Pre->Post)       | Pts |  1  |  2  |  3  |  4  |  5  |  6  |  7  | '],
 ['-----------------------------------------------------------------------------------------'],
 ['    1 | GARY HUA                        |6.0  |W  39|W  21|W  18|W  14|W   7|D  12|D   4|'],
 ['   ON | 15445895 / R: 1794   ->1817     |N:2  |W    |B    |W    |B    |W    |B    |W    |'],
 ['-----------------------------------------------------------------------------------------'],
 ['    2 | DAKSHESH DARURI                 |6.0  |W  63|W  58|L   4|W  17|W  16|W  20|W   7|'],
 ['   MI | 14598900 / R: 1553   ->1663     |N:2  |B    |W    |B    |W    |B    |W    |B    |'],
 ['-----------------------------------------------------------------------------------------'],
 ['    3 | ADITYA BAJAJ               

Use "a = a[3:]" to remove first three rows, arange 'a' to two lists, row1 and row2, row1 is 3 times length+2, strat at 2,step 3 for each row and end at the length of a ,row2 is 3 times length +3, start at 3, step 3 for each row and also end at the length of a

In [228]:
# remove first three rows
a = a[3:]
# use len function to get the length of a
length = len(a)
# use np.arange to arange a, start at 2,step 3, end at 3
row1_num = np.arange(1,length,3)
step = 0
# create a empty list to store first arange list
row1 = []
#create a for loop, if variable step in row1_num, store that row to row1[]
for i in a:
    if(step in row1_num):
        row1.append(i)
    step = step+1
#print row1 
row1

[['    1 | GARY HUA                        |6.0  |W  39|W  21|W  18|W  14|W   7|D  12|D   4|'],
 ['    2 | DAKSHESH DARURI                 |6.0  |W  63|W  58|L   4|W  17|W  16|W  20|W   7|'],
 ['    3 | ADITYA BAJAJ                    |6.0  |L   8|W  61|W  25|W  21|W  11|W  13|W  12|'],
 ['    4 | PATRICK H SCHILLING             |5.5  |W  23|D  28|W   2|W  26|D   5|W  19|D   1|'],
 ['    5 | HANSHI ZUO                      |5.5  |W  45|W  37|D  12|D  13|D   4|W  14|W  17|'],
 ['    6 | HANSEN SONG                     |5.0  |W  34|D  29|L  11|W  35|D  10|W  27|W  21|'],
 ['    7 | GARY DEE SWATHELL               |5.0  |W  57|W  46|W  13|W  11|L   1|W   9|L   2|'],
 ['    8 | EZEKIEL HOUGHTON                |5.0  |W   3|W  32|L  14|L   9|W  47|W  28|W  19|'],
 ['    9 | STEFANO LEE                     |5.0  |W  25|L  18|W  59|W   8|W  26|L   7|W  20|'],
 ['   10 | ANVIT RAO                       |5.0  |D  16|L  19|W  55|W  31|D   6|W  25|W  18|'],
 ['   11 | CAMERON WILLIAM MC LEMAN     

In [225]:
# use len function to get the length of a
length = len(a)
# use np.arange to arange a, start at 3,step 3, end at 3
row2_num = np.arange(2,length,3)
step = 0
# create a empty list to store first arange list
row2 = []
#create a for loop, if variable step in row2_num, store that row to row2[]
for i in a:
    if(step in row2_num):
        row2.append(i)
    step = step+1
#print row2
row2

[['   ON | 15445895 / R: 1794   ->1817     |N:2  |W    |B    |W    |B    |W    |B    |W    |'],
 ['   MI | 14598900 / R: 1553   ->1663     |N:2  |B    |W    |B    |W    |B    |W    |B    |'],
 ['   MI | 14959604 / R: 1384   ->1640     |N:2  |W    |B    |W    |B    |W    |B    |W    |'],
 ['   MI | 12616049 / R: 1716   ->1744     |N:2  |W    |B    |W    |B    |W    |B    |B    |'],
 ['   MI | 14601533 / R: 1655   ->1690     |N:2  |B    |W    |B    |W    |B    |W    |B    |'],
 ['   OH | 15055204 / R: 1686   ->1687     |N:3  |W    |B    |W    |B    |B    |W    |B    |'],
 ['   MI | 11146376 / R: 1649   ->1673     |N:3  |W    |B    |W    |B    |B    |W    |W    |'],
 ['   MI | 15142253 / R: 1641P17->1657P24  |N:3  |B    |W    |B    |W    |B    |W    |W    |'],
 ['   ON | 14954524 / R: 1411   ->1564     |N:2  |W    |B    |W    |B    |W    |B    |B    |'],
 ['   MI | 14150362 / R: 1365   ->1544     |N:3  |W    |W    |B    |B    |W    |B    |W    |'],
 ['   MI | 12581589 / R: 1712   ->1696  

create a regular expression pattern and store in b, and use findall function to get all of the name in first arange list 'row1'

In [102]:
# create a regular expression pattern and store in b
b = re.compile("[a-zA-Z]+\s[a-zA-Z]+\s[a-zA-Z]*")
# use findall function find all values which match the pattern
name = b.findall(str(row1))
#print name
name

['GARY HUA ',
 'DAKSHESH DARURI ',
 'ADITYA BAJAJ ',
 'PATRICK H SCHILLING',
 'HANSHI ZUO ',
 'HANSEN SONG ',
 'GARY DEE SWATHELL',
 'EZEKIEL HOUGHTON ',
 'STEFANO LEE ',
 'ANVIT RAO ',
 'CAMERON WILLIAM MC',
 'KENNETH J TACK',
 'TORRANCE HENRY JR',
 'BRADLEY SHAW ',
 'ZACHARY JAMES HOUGHTON',
 'MIKE NIKITIN ',
 'RONALD GRZEGORCZYK ',
 'DAVID SUNDEEN ',
 'DIPANKAR ROY ',
 'JASON ZHENG ',
 'DINH DANG BUI',
 'EUGENE L MCCLURE',
 'ALAN BUI ',
 'MICHAEL R ALDRICH',
 'LOREN SCHWIEBERT ',
 'MAX ZHU ',
 'GAURAV GIDWANI ',
 'SOFIA ADINA STANESCU',
 'CHIEDOZIE OKORIE ',
 'GEORGE AVERY JONES',
 'RISHI SHETTY ',
 'JOSHUA PHILIP MATHEWS',
 'JADE GE ',
 'MICHAEL JEFFERY THOMAS',
 'JOSHUA DAVID LEE',
 'SIDDHARTH JHA ',
 'AMIYATOSH PWNANANDAM ',
 'BRIAN LIU ',
 'JOEL R HENDON',
 'FOREST ZHANG ',
 'KYLE WILLIAM MURPHY',
 'JARED GE ',
 'ROBERT GLEN VASEY',
 'JUSTIN D SCHILLING',
 'DEREK YAN ',
 'JACOB ALEXANDER LAVALLEY',
 'ERIC WRIGHT ',
 'DANIEL KHAIN ',
 'MICHAEL J MARTIN',
 'SHIVAM JHA ',
 'TEJAS A

create a regular expression pattern and store in b, and use findall function to get all of the Pre_Rating in first arange list 'row1'

In [141]:
# create a regular expression pattern and store in b
b = re.compile("[0-9]\.[0-9]")
# use findall function find all values which match the pattern
Pre_Rating = b.findall(str(row1))
# print Pre_Rating
Pre_Rating

['6.0',
 '6.0',
 '6.0',
 '5.5',
 '5.5',
 '5.0',
 '5.0',
 '5.0',
 '5.0',
 '5.0',
 '4.5',
 '4.5',
 '4.5',
 '4.5',
 '4.5',
 '4.0',
 '4.0',
 '4.0',
 '4.0',
 '4.0',
 '4.0',
 '4.0',
 '4.0',
 '4.0',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.5',
 '3.0',
 '3.0',
 '3.0',
 '3.0',
 '3.0',
 '3.0',
 '3.0',
 '3.0',
 '3.0',
 '2.5',
 '2.5',
 '2.5',
 '2.5',
 '2.5',
 '2.5',
 '2.0',
 '2.0',
 '2.0',
 '2.0',
 '2.0',
 '2.0',
 '2.0',
 '1.5',
 '1.5',
 '1.0',
 '1.0',
 '1.0']

Create a regular expression pattern and store in b, and use findall function to get all of the state in second arange list 'row2'

In [230]:
# create a regular expression pattern and store in b
b = re.compile("[a-zA-Z]{2}")
# use findall function find all values which match the pattern
state = b.findall(str(row2))
# print state
state

['ON',
 'MI',
 'MI',
 'MI',
 'MI',
 'OH',
 'MI',
 'MI',
 'ON',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'ON',
 'MI',
 'ON',
 'MI',
 'MI',
 'ON',
 'MI',
 'MI',
 'MI',
 'ON',
 'MI',
 'ON',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'MI',
 'ON',
 'MI',
 'MI',
 'MI']

create a regular expression pattern and store in b, but the result would like 'R: 1794', so I create a second regular expression pattern to remove R: from the first result,

In [233]:
# create a regular expression pattern and store in b, 
b = re.compile("[A-Z]{1}\:\s+[0-9]{3,4}")
# use findall function find all values which match the pattern
points = b.findall(str(row2))
# create a new regular expression pattern and store in bb,
bb = re.compile("[0-9]+")
# use findall function find all values which match the pattern
points = bb.findall(str(points))       
# print points
points

['1794',
 '1553',
 '1384',
 '1716',
 '1655',
 '1686',
 '1649',
 '1641',
 '1411',
 '1365',
 '1712',
 '1663',
 '1666',
 '1610',
 '1220',
 '1604',
 '1629',
 '1600',
 '1564',
 '1595',
 '1563',
 '1555',
 '1363',
 '1229',
 '1745',
 '1579',
 '1552',
 '1507',
 '1602',
 '1522',
 '1494',
 '1441',
 '1449',
 '1399',
 '1438',
 '1355',
 '980',
 '1423',
 '1436',
 '1348',
 '1403',
 '1332',
 '1283',
 '1199',
 '1242',
 '377',
 '1362',
 '1382',
 '1291',
 '1056',
 '1011',
 '935',
 '1393',
 '1270',
 '1186',
 '1153',
 '1092',
 '917',
 '853',
 '967',
 '955',
 '1530',
 '1175',
 '1163']

Firstly create a while loop, if variable i less than the length of row1, create a regular expression pattern and store in b, but the result would like "39|", so I create a second regular expression pattern, to remove | from the result, and get the pair number of each rows.
After get the pair number, we created a for loop, we use for loop to get the each element from that row, and find the total points from that row, and divide by length of num, to get the Ave_Rating gor each rows

In [201]:
# create two variable i and temp
i = 0
temp = 0
# create a empty list to store value
Avg_Rating = []
# create a while loop
while(i<len(row1)):
    # create a regular expression pattern and store in b
    b = re.compile("[0-9]+\|")
    # use findall function find all values which match the pattern
    num = b.findall(str(row1[i]))
    # create a regular expression pattern and store in bb
    bb = re.compile("[0-9]+")
    # use findall function find all values which match the pattern
    num = bb.findall(str(num))
    #create a for loop, gather total points for each row
    for x in range(len(num)):
        temp = temp + int(points[int(num[x])-1])
    # get the Avg_Rating for each row and use round function to round to 2 decimal                  
    Avg_Rating.append(round(temp/len(num),2))
      
    i=i+1
    temp = 0
#pirnt result
Avg_Rating

[1605.29,
 1469.29,
 1563.57,
 1573.57,
 1500.86,
 1518.71,
 1372.14,
 1468.43,
 1523.14,
 1554.14,
 1467.57,
 1506.17,
 1497.86,
 1515.0,
 1483.86,
 1385.8,
 1498.57,
 1480.0,
 1426.29,
 1410.86,
 1470.43,
 1300.33,
 1213.86,
 1357.0,
 1363.29,
 1506.86,
 1221.67,
 1522.14,
 1313.5,
 1144.14,
 1259.86,
 1378.71,
 1276.86,
 1375.29,
 1149.71,
 1388.17,
 1384.8,
 1539.17,
 1429.57,
 1390.57,
 1248.5,
 1149.86,
 1106.57,
 1327.0,
 1152.0,
 1357.71,
 1392.0,
 1355.8,
 1285.8,
 1296.0,
 1356.14,
 1494.57,
 1345.33,
 1206.17,
 1406.0,
 1414.4,
 1363.0,
 1391.0,
 1319.0,
 1330.2,
 1327.29,
 1186.0,
 1350.2,
 1263.0]

Use the data from above to create a dataframe

In [236]:
data = {'Name':name,'State':state,'Pre_Rating':Pre_Rating,'Points':points,'Avg_Rating':Avg_Rating} 
# Create DataFrame 
df = pd.DataFrame(data) 
df

Name State Pre_Rating Points  Avg_Rating
0                  GARY HUA     ON        6.0   1794     1605.29
1           DAKSHESH DARURI     MI        6.0   1553     1469.29
2              ADITYA BAJAJ     MI        6.0   1384     1563.57
3        PATRICK H SCHILLING    MI        5.5   1716     1573.57
4                HANSHI ZUO     MI        5.5   1655     1500.86
5               HANSEN SONG     OH        5.0   1686     1518.71
6          GARY DEE SWATHELL    MI        5.0   1649     1372.14
7          EZEKIEL HOUGHTON     MI        5.0   1641     1468.43
8               STEFANO LEE     ON        5.0   1411     1523.14
9                 ANVIT RAO     MI        5.0   1365     1554.14
10        CAMERON WILLIAM MC    MI        4.5   1712     1467.57
11            KENNETH J TACK    MI        4.5   1663     1506.17
12         TORRANCE HENRY JR    MI        4.5   1666     1497.86
13             BRADLEY SHAW     MI        4.5   1610     1515.00
14    ZACHARY JAMES HOUGHTON    MI        4.5   1220     1483.86
15             MIKE NIKITIN     MI        4.0   1604     1385.80
16       RONALD GRZEGORCZYK     MI        4.0   1629     1498.57
17            DAVID SUNDEEN     MI        4.0   1600     1480.00
18             DIPANKAR ROY     MI        4.0   1564     1426.29
19              JASON ZHENG     MI        4.0   1595     1410.86
20             DINH DANG BUI    ON        4.0   1563     1470.43
21          EUGENE L MCCLURE    MI        4.0   1555     1300.33
22                 ALAN BUI     ON        4.0   1363     1213.86
23         MICHAEL R ALDRICH    MI        4.0   1229     1357.00
24         LOREN SCHWIEBERT     MI        3.5   1745     1363.29
25                  MAX ZHU     ON        3.5   1579     1506.86
26           GAURAV GIDWANI     MI        3.5   1552     1221.67
27      SOFIA ADINA STANESCU    MI        3.5   1507     1522.14
28         CHIEDOZIE OKORIE     MI        3.5   1602     1313.50
29        GEORGE AVERY JONES    ON        3.5   1522     1144.14
..                       ...   ...        ...    ...         ...
34          JOSHUA DAVID LEE    MI        3.5   1438     1149.71
35            SIDDHARTH JHA     MI        3.5   1355     1388.17
36     AMIYATOSH PWNANANDAM     MI        3.5    980     1384.80
37                BRIAN LIU     MI        3.0   1423     1539.17
38             JOEL R HENDON    MI        3.0   1436     1429.57
39             FOREST ZHANG     MI        3.0   1348     1390.57
40       KYLE WILLIAM MURPHY    MI        3.0   1403     1248.50
41                 JARED GE     MI        3.0   1332     1149.86
42         ROBERT GLEN VASEY    MI        3.0   1283     1106.57
43        JUSTIN D SCHILLING    MI        3.0   1199     1327.00
44                DEREK YAN     MI        3.0   1242     1152.00
45  JACOB ALEXANDER LAVALLEY    MI        3.0    377     1357.71
46              ERIC WRIGHT     MI        2.5   1362     1392.00
47             DANIEL KHAIN     MI        2.5   1382     1355.80
48          MICHAEL J MARTIN    MI        2.5   1291     1285.80
49               SHIVAM JHA     MI        2.5   1056     1296.00
50           TEJAS AYYAGARI     MI        2.5   1011     1356.14
51                ETHAN GUO     MI        2.5    935     1494.57
52             JOSE C YBARRA    MI        2.0   1393     1345.33
53              LARRY HODGE     MI        2.0   1270     1206.17
54                ALEX KONG     MI        2.0   1186     1406.00
55             MARISA RICCI     MI        2.0   1153     1414.40
56               MICHAEL LU     MI        2.0   1092     1363.00
57             VIRAJ MOHILE     MI        2.0    917     1391.00
58                 SEAN M MC    MI        2.0    853     1319.00
59               JULIA SHEN     MI        1.5    967     1330.20
60            JEZZEL FARKAS     ON        1.5    955     1327.29
61            ASHWIN BALAJI     MI        1.0   1530     1186.00
62      THOMAS JOSEPH HOSMER    MI        1.0   1175     1350.20
63                   BEN LI     MI        1.0   1163     12

convert dataframe to csv file, name data.csv

In [237]:
df.to_csv("data.csv")